docs/treeview_performance.md

```diff
@@ -0,0 +1,136 @@
+
+Some alternatives we can adopt here to solve this is virtualization as recommended by the React.js team which will only render a limited amount of Nodes according to the viewport.
+
+#### Expand a node
+
+<img alt="Tracing expand a node" src="https://user-images.githubusercontent.com/13750819/140421389-a2098c37-5e5d-42c0-a9a9-4d54839b67ed.png" style="border-radius: 5px;" />
+
+The API decision to avoid tree traversal to change a property on each Node that would symbolize whether the Node is expanded or not also has the same impact as causing a render on all other Nodes but avoids the cost of traversing the tree.
+
+Since all Nodes consume the same context to know if the Node is expanded any new items added to the state will cause it to re-render.
+
+Ideally, just cause a rendering on the Node that is being expanded, we have some alternatives to avoid that like treating the `expandedKeys` only as initial and using local state on the Node to expand, we can have some drastic performance gains.
+
+This is a common problem for context usage that all-consuming components will cause re-rendering, there is a proposal for this in [React RFC 119 (Context selectors)](https://github.com/reactjs/rfcs/pull/119) that would render the component only if the data it is consuming has changed, a [PR from the React team](https://github.com/facebook/react/pull/20646) was sent some time ago as an experiment but still open.
+
+#### Select a node
+
+<img alt="Tracing select a node" src="https://user-images.githubusercontent.com/13750819/140422167-4cb42107-f11c-41b9-8079-f1d1f261fa80.png" style="border-radius: 5px;" />
+
+Selecting a node is similar to expanding because it has the same API design, we can use the same alternatives to avoid rendering on all nodes.
+
+### Drag and Drop
+
+| Dragging an item | Dropping an item |
+| ---------------- | ---------------- |
+| 5025ms / 3 ops   | 5930ms / 2 ops   |
+
+The implementation of DnD is built on top of the [react-dnd](https://react-dnd.github.io/) library which we ended up depending on the performance of the library but we made our improvements that make the tree change without a high cost, it's down to about a few nanoseconds with the immutability of the tree.
+
+The perception of slowness is totally noticeable to start dragging and then being able to move the Node between the elements is very slow generating animation slowdowns and the drop also has the same effect. The reason is very similar to Select a node and Expand, `react-dnd` causes a lot of rendering in its hooks which are used on all Nodes which makes all components render in one big list takes up a lot of thread time, a list is not phased so it takes a lot of main thread time.
```

Comment:
I can't think of a way to smooth DnD for these large data, we rely heavily on react-dnd for that.